File: DP/2dDP/knapsack.cpp
#include<iostream>
#include<vector>
using namespace std;    


// top down dp
int findMaxProfit(int n,int wt,vector<int>&wts,vector<int>&prices,vector<vector<int>>&dp){
    //base case
    if(n==0 or wt==0){
        return 0;
    }
    if(dp[n][wt]!=0){
        return dp[n][wt];
    }
    int inc=0;
    if(wt-wts[n-1]>=0){
        inc=prices[n-1]+findMaxProfit(n-1,wt-wts[n-1],wts,prices,dp);
    }
    
    int exc=findMaxProfit(n-1,wt,wts,prices,dp);
    return dp[n][wt]=max(inc,exc);
}


// bottom-up dp
int findMaxProfit(int n,int wt,vector<int>&wts,vector<int>prices){
    vector<vector<int>>dp(n+1,vector<int>(wt+1,0)); //stores the profit for i items and j capicity of the bag
    for(int i=1;i<=n;i++){
        for(int j=1;j<=wt;j++){
            int inc=0,exc=0;
            if(wts[i-1]<=j){
                inc+=prices[i]+dp[i-1][j-wts[i-1]];
            }
            exc+=dp[i-1][j];
            dp[i][j]=max(inc,exc);
        }
    }

    return dp[n][wt];
}






int main(){
    int n=4;
    int wt=11;
    vector<int>wts{2,7,3,4};
    vector<int>prices{5,20,20,10};


    // vector<vector<int>>dp(n+1,vector<int>(wt+1,0));
    // cout<<findMaxProfit(n,wt,wts,prices,dp)<<endl;  //f(n,w) denotes the maxProfit for n items and and bag with weight wt

    cout<<findMaxProfit(n,wt,wts,prices)<<endl;


}

File: DP/2dDP/knapsack.py
# top down dp
def findMaxProfit(i,max_w,wts,prices,dp):
    if i==0:
        if wts[i]<=max_w:
            return prices[0]
        else:
            return 0

    if dp[i][max_w]!=-1:
        return dp[i][max_w]

    exc=findMaxProfit(i-1,max_w,wts,prices,dp)
    inc=0
    if wts[i]<=max_w:
        inc=prices[i]+findMaxProfit(i-1,max_w-wts[i],wts,prices,dp)

    dp[i][max_w]=max(exc,inc)
    return dp[i][max_w]


def findMaxProfit(n,max_w,wts,prices):
    # dp=[[0]*(max_w+1) for _ in range(n)]
    #space optimization
    prev=[0]*(max_w+1)
    curr=[0]*(max_w+1)

    # base case
    for wt in range(wts[0],max_w+1):
        # dp[0][wt]=prices[0]
        prev[wt]=prices[0]

    for i in range(1,n):
        for wt in range(1,max_w+1):
            # exc=dp[i-1][wt]
            exc=prev[wt]
            inc=0
            if wts[i]<=wt:
                # inc=prices[i]+dp[i-1][wt-wts[i]]
                inc=prices[i]+prev[wt]

            # dp[i][wt]=max(exc,inc)
            curr[wt]=max(exc,inc)

        prev=curr


    # return dp[n-1][max_w]
    return prev[max_w]

t=int(input())
for _ in range(t):
    n,max_w=list(map(int,input().split()))
    wts=list(map(int,input().split()))
    prices=list(map(int,input().split()))


    # dp=[[-1]*(max_w+1) for _ in range(n)]

    # print(findMaxProfit(n-1,max_w,wts,prices,dp))


    print(findMaxProfit(n,max_w,wts,prices))
